File: 总题库/227.BasicCalculatorII.py
class Solution:
    def calculate(self, s: str) -> int:
        st = []
        res, index, sign, sLen= 0, 0, 1, len(s)

        while index < sLen:
            if s[index].isdigit():
                num = 0
                while index < sLen and s[index].isdigit():
                    num = 10 * num + int(s[index])
                    index += 1
                index -= 1
                st.append(sign * num)
            else:
                if s[index] == '-': sign = -1
                if s[index] == '+': sign = 1
                if s[index] == '*' or s[index] == '/':
                    num1, num2, num3= st[-1], 0, 0
                    st.pop()
                    opt = s[index]
                    index += 1
                    if index == sLen: return
                    while s[index] == ' ': index += 1
                    while index < sLen and s[index].isdigit():
                        num2 = num2 * 10 + int(s[index])
                        index += 1
                    index -= 1
                    if opt == '/':
                        num3 = int(num1 / num2)
                    elif opt == '*':
                        num3 = int(num1 * num2)
                    st.append(num3)
            index += 1

        for i in st:
            res += i
        return res

File: 总题库/227.BasicCalculatorII.cpp
class Solution {
public:
    int calculate(string s) {
        stack<int> st;
        int sign = 1, sLen = s.size(), res = 0;
        for (int i = 0; i < sLen; ++i) {
            if (isdigit(s[i])) {
                int num = 0;
                while (i < sLen && isdigit(s[i])) {
                    num = 10 * num + s[i] - '0';
                    i++;
                }
                i--;
                st.push(sign * num);
            }
            else {
                if (s[i] == '+') sign = 1;
                if (s[i] == '-') sign = -1;
                if (s[i] == '/' || s[i] == '*') {
                    char opt = s[i];
                    int num1 = st.top(), num2 = 0, num3 = 0;
                    st.pop();
                    i++;
                    if (i == sLen) return 0; // 输入出错
                    while (i < sLen && (isdigit(s[i]) || s[i] == ' ')) {
                        num2 = 10 * num2 + s[i] - '0';
                        i++;
                    }
                    i--;
                    if (opt == '/') {
                        num3 = num1 / num2;
                    }
                    else if (opt == '*') {
                        num3 = num1 * num2;
                    }
                    st.push(num3);
                }
            }
        }
        while (!st.empty()) {
            res += st.top();
            st.pop();
        }
        return res;
    }
};

// "3+2*2" = 7
// " 3/2 " = 1
// " 3+5 / 2 " = 5